README.md
[![Build Status](https://travis-ci.org/F5Networks/cf-bigip-ctlr.svg?branch=master)](https://travis-ci.org/F5Networks/cf-bigip-ctlr) [![Slack](https://f5cloudsolutions.herokuapp.com/badge.svg)](https://f5cloudsolutions.herokuapp.com) [![Coverage Status](https://coveralls.io/repos/github/F5Networks/cf-bigip-ctlr/badge.svg?branch=HEAD)](https://coveralls.io/github/F5Networks/cf-bigip-ctlr?branch=HEAD)

F5 BIG-IP Controller for Cloud Foundry
======================================

The F5 BIG-IP Controller for [Cloud Foundry](http://cloudfoundry.org) makes the F5 BIG-IP
[Local Traffic Manager](<https://f5.com/products/big-ip/local-traffic-manager-ltm)
services available to applications running in the Cloud Foundry platform.

Documentation
-------------

For instructions on how to use this component, use the
[F5 BIG-IP Controller for Cloud Foundry docs](http://clouddocs.f5.com/products/connectors/cf-bigip-ctlr/latest/).

For guides on this and other solutions for Cloud Foundry, see the
[F5 Solution Guides for Cloud Foundry](http://clouddocs.f5.com/containers/latest/cloudfoundry).

Getting Help
------------

We encourage you to use the cc-cloudfoundry channel in our [f5CloudSolutions Slack workspace](https://f5cloudsolutions.herokuapp.com/) for discussion and assistance on this
controller. This channel is typically monitored Monday-Friday 9am-5pm MST by F5
employees who will offer best-effort support.

Contact F5 Technical support via your typical method for more time sensitive
changes and other issues requiring immediate support.

Running
-------

The official docker image is `f5networks/cf-bigip-ctlr`.

Usually, the controller is deployed in Cloud Foundry. However, the controller can be run locally for development testing.
The controller requires a running NATS server, without a valid connection the controller will not start. The controller
can either be run against a gnatsd server or standalone against a Cloud Foundry installation.

Option gnatsd:
- Go should be installed and in the PATH
- GOPATH should be set as described in http://golang.org/doc/code.html
- Pip install the python/cf-runtime-requirements.txt into a virtualenv of your choice
```bash
workon cf-bigip-ctlr
pip install -r python/cf-runtime-requirements.txt
```
- [gnatds](https://github.com/nats-io/gnatds) installed and in the PATH
```
go get github.com/nats-io/gnatsd
gnatsd &
```
- Optionally run unit tests
```
go get github.com/onsi/ginkgo
go get github.com/onsi/gomega
ginkgo -keepGoing -trace -p -progress -r -failOnPending -randomizeAllSpecs -race
```
- Build and install controller from a cloned [cf-bigip-ctlr](https://github.com/F5Networks/cf-bigip-ctlr)
```
go install
```
- Update BIGIP_CTLR_CFG environment variable for your specific environment
  as described in "Configuration"
- Run the controller
```
cf-bigip-ctlr
```

Option standalone:
- Go should be installed and in the PATH
- GOPATH should be set as described in http://golang.org/doc/code.html
- Pip install the python/cf-runtime-requirements.txt into a virtualenv of your choice
```bash
workon cf-bigip-ctlr
pip install -r python/cf-runtime-requirements.txt
```
- Optionally run unit tests
```
go get github.com/onsi/ginkgo
go get github.com/onsi/gomega
ginkgo -keepGoing -trace -p -progress -r -failOnPending -randomizeAllSpecs -race
```
- Build and install controller from a cloned [cf-bigip-ctlr](https://github.com/F5Networks/cf-bigip-ctlr)
```
go install
```
- Update configuration file or BIGIP_CTLR_CFG environment variable for your specific environment
  as described in "Configuration"
- Run the controller
```
cf-bigip-ctlr -c [CONFIG_FILE]
```

Building
--------

The official images are built using docker, but standard go build tools can be used for development
purposes as described above.

### Official Build

Prerequisites:
- Docker

```bash
git clone https://github.com/F5Networks/cf-bigip-ctlr.git
cd cf-bigip-ctlr

# Use docker to build the release artifacts into a local "_docker_workspace" directory and push into docker images
make prod
```

### Alternate, unofficial build

A normal go and godep toolchain can be used as well

Prerequisites:
- go 1.7
- GOPATH pointing at a valid go workspace
- godep (Only needed to modify vendor's packages)
- python
- virtualenv

```bash
mkdir -p $GOPATH/src/github.com/F5Networks
cd $GOPATH/src/github.com/F5Networks
git clone https://github.com/F5Networks/cf-bigip-ctlr.git
cd cf-bigip-ctlr

# Building all packages, and run unit tests
make prod
```

Configuration
-------------

When pushing the controller into a Cloud Foundry environment a configuration must be passed
via the application manifest. An example manifest is located in the
`docs/_static/config_examples` directory.

Update required sections for environment:
- nats: leave empty for gnatsd otherwise update with CF installed NATS information
- bigip: leave empty if no BIG-IP is required otherwise update with BIG-IP information
- routing_api: only required if routing API access is required (TCP routing with route_mode
set to all, or tcp)
- oauth: only required if routing API access is required (TCP routing with route_mode
set to all, or tcp)

On startup, the controller will get the `BIGIP_CTLR_CFG` variable from the environment;
setting this variable via an application manifest is required, environment parameters
are set in the manifest `env` section. The controller also supports Cloud Foundry health
checks. These can be configured in the manifest fields: `health-check-type`, and
`health-check-http-endpoint`. In order to configure health checking of the controller use
these settings:

```
health-check-type: http
health-check-http-endpoint: /health
```

To explore what other settings are available refer to the Cloud Foundry documentation
[Deploying with Application Manifests](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html).

A minimal configuration manifest to support HTTP routing mode would be written as such
(`route_mode` set to 'http'):
```
applications:
  - name: cf-bigip-ctlr
    health-check-type: http
    health-check-http-endpoint: /health
    env:
      BIGIP_CTLR_CFG: |
                      bigip:
                        url: https://bigip.example.com
                        user: admin
                        pass: password
                        partition:
                          - example
                        external_addr: 192.168.1.1
                      nats:
                        - host: 192.168.10.1
                          port: 4222
                          user: nats
                          pass: nats-password
                      route_mode: http
```

If TCP routing is required, the `oauth` and `routing_api` sections are required
but not the `nats` section. Set `route_mode` to 'tcp'.

If both modes are required, the `oauth`, `routing_api`, and `nats` sections are required.
Set `route_mode` to 'all'.

Development
-----------

**Note**: This repository should be imported as `github.com/F5Networks/cf-bigip-ctlr`.

## Dynamic Routing Table

The controller's routing table is updated dynamically via the NATS message bus.
NATS can be deployed via BOSH with
([cf-release](https://github.com/cloudfoundry/cf-release)) or standalone using
[nats-release](https://github.com/cloudfoundry/nats-release).

To add or remove a record from the routing table, a NATS client must send
register or unregister messages. Records in the routing table have a maximum
TTL of 120 seconds, so clients must heartbeat registration messages
periodically; we recommend every 20s. [Route
Registrar](https://github.com/cloudfoundry/route-registrar) is a BOSH job that
comes with [Routing
Release](https://github.com/cloudfoundry-incubator/routing-release) that
automates this process.

When deployed with Cloud Foundry, registration of routes for apps pushed to CF
occurs automatically without user involvement. For details, see [Routes and
Domains](https://docs.cloudfoundry.org/devguide/deploy-apps/routes-domains.html).

### Registering Routes via NATS

When the controller starts, it sends a `router.start` message to NATS. This
message contains an interval that other components should then send
`router.register` on, `minimumRegisterIntervalInSeconds`. It is recommended
that clients should send `router.register` messages on this interval. This
`minimumRegisterIntervalInSeconds` value is configured through the
`start_response_delay_interval` configuration property. The controller will prune
routes that it considers to be stale based upon a separate "staleness" value,
`droplet_stale_threshold`, which defaults to 120 seconds. The controller will check
if routes have become stale on an interval defined by
  `prune_stale_droplets_interval`, which defaults to 30 seconds. All of these
  values are represented in seconds and will always be integers.

The format of the `router.start` message is as follows:

```json
{
  "id": "some-router-id",
  "hosts": ["1.2.3.4"],
  "minimumRegisterIntervalInSeconds": 20,
  "prunteThresholdInSeconds": 120,
}
```

After a `router.start` message is received by a client, the client should send
`router.register` messages. This ensures that the new controller can update its
routing table.

If a component comes online after the controller, it must make a NATS request
called `router.greet` in order to determine the interval. The response to this
message will be the same format as `router.start`.

The format of the `router.register` message is as follows:

```json
{
  "host": "127.0.0.1",
  "port": 4567,
  "uris": [
    "my_first_url.vcap.me",
    "my_second_url.vcap.me"
  ],
  "tags": {
    "another_key": "another_value",
    "some_key": "some_value"
  },
  "app": "some_app_guid",
  "stale_threshold_in_seconds": 120,
  "private_instance_id": "some_app_instance_id",
  "router_group_guid": "some_router_group_guid"
}
```

`stale_threshold_in_seconds` is the custom staleness threshold for the route
being registered. If this value is not sent, it will default to the controller's
default staleness threshold.

`app` is a unique identifier for an application that the endpoint is registered
for.

`private_instance_id` is a unique identifier for an instance associated with
the app identified by the `app` field.

`router_group_guid` determines which controllers will register route. Only
controllers configured with the matching router group will register the route. If
a value is not provided, the route will be registered by all controllers that
have not be configured with a router group.

Such a message can be sent to both the `router.register` subject to register
URIs, and to the `router.unregister` subject to unregister URIs, respectively.

**Note:** In order to use `nats-pub` to register a route, you must run the command on the NATS VM. If you are using [`cf-deployment`](https://github.com/cloudfoundry/cf-deployment), you can run `nats-pub` from any VM.


docs/RELEASE-NOTES.rst
Release Notes for BIG-IP Controller for Cloud Foundry
=====================================================

v1.2.1
-----

Bug Fixes
`````````
* :issues:`171` - Controller upgrades properly with new metadata field.

v1.2.0
------

Added Functionality
```````````````````
* Added controller name and version to the metadata of certain BIG-IP LTM resources managed by the controller.
* Base container image upgraded from Alpine 3.4 to Alpine 3.7.

Bug Fixes
`````````
* :issues:`150` - Fix go vet lock copy error
* :issues:`134` - Add tier2_ip_range validation more robust to match BIG-IP input requirements.

Limitations
```````````
* If an older controller created resources, upgrading to the new version could result in a python
  exception when adding metadata to virtuals: :issues:`171`

v1.1.1
------

Bug Fixes
`````````
* :cccl-issue:`208` - Address compatibility for BIG-IP v13.0 Health Monitor interval and timeout.
* :cccl-issue:`211` - Fixed memory leak in python subprocess.
* :issues:`127` - Make tier2_ip_range parameter optional (to be backwards compatible)
* :issues:`113` - Change behavior of controller (do not exit) on startup when encountering errors 
* :issues:`112` - Controller should not stop updating BIG-IP configuration after an exception occurs in the python driver

v1.1.0
------

Added Functionality
```````````````````
* Use the Controller as a Service Broker to apply per-route L7 (HTTP) configurations.

  * Virtual Server for L7 (HTTP) route can have its own Policies, Profiles and SSL Profiles.
  * Pool for L7 (HTTP) route can have its own load balancing mode and Health Monitors.

* Adopts a new, two-tier architecture. See the BIG-IP Controller for Cloud Foundry `user documentation`_ for more information.
* Support for JSESSIONID cookie session persistence.

Limitations
```````````
* Architected only as a Static, Brokered Service.
* Controller doesn't support Service bindings.
* Controller cannot accept Arbitrary Parameters via ``cf create-service|bind-service -c``.
* Controller doesn't support the use of ``cf update-service``.
* The BIG-IP data group cf-ctlr-data-group, found in the same partition that the controller manages, needs verification if the controller crashes. There should be a one to one mapping of tier 2 VIPs to data group records with the same name. If the record is missing for a VIP the controller will not rebuild the BIG-IP to the correct state it was in before the crash. This is because the controller crashed before the updated data group got written out. This condition requires manual intervention. Update the data group with the missing VIP's name as key and virtual address and port as data. The data must follow this format: {"bindAddr":"ADDRESS","port":PORT}. Encode the data as base64.
* JSESSIONID cookie session persistence will not work for BIG-IP v11.6.1 and will cause `Connection Reset by Peer` connection errors. Correct this by disabling session persistence, set session_persistence to false in the controller config.


v1.0.0
------

Added Functionality
```````````````````
* Support for JSESSIONID cookie session persistence.
* Support for TCP and HTTP routing.
* Attach custom policy, profile, or health monitor to L7 objects created on the BIG-IP device.
* Manages the following Local Traffic Manager (LTM) resources for the BIG-IP partition:

  * Virtual Servers
  * Pools
  * Pool members
  * Nodes
  * Policies

    * Rules

      * Actions
      * Conditions

Limitations
```````````
* The BIG-IP Controller controls one (1) partition on the BIG-IP device.
* Controller configurations are global: they apply to all L7 (HTTP) LTM objects in the designated BIG-IP partition.
* This release supports custom policies and profiles for **L7 virtual servers** only.
* Configured health monitor objects apply to all pools (both L4 and L7 routes).
* SSL profile(s) defined in the application manifest do not attach to the HTTP virtual server.
* Modification of a Controller-owned policy resulting in a state change may cause traffic flow interruptions. If the modification changes the state to ‘published’, the Controller will delete the policy and recreate it with a ‘legacy’ status.
* You cannot change the default route domain for a partition managed by an F5 controller after the controller has deployed. To specify a new default route domain, use a different partition.


config/config.go
/*
 * Portions Copyright (c) 2017,2018, F5 Networks, Inc.
 */

package config

import (
	"crypto/tls"
	"fmt"
	"io/ioutil"
	"net/url"
	"runtime"
	"strings"
	"time"

	"code.cloudfoundry.org/localip"
	"gopkg.in/yaml.v2"
)

// RoutingMode of controller
type RoutingMode int

const (
	// TCP only routing mode
	TCP RoutingMode = iota
	// HTTP only routing mode
	HTTP
	// all for TCP and HTTP
	all
)

func (rm RoutingMode) String() string {
	switch rm {
	case TCP:
		return "tcp"
	case HTTP:
		return "http"
	case all:
		return "all"
	}
	return "Unknown"
}

const (
	LOAD_BALANCE_RR string = "round-robin"
	LOAD_BALANCE_LC string = "least-connection"
)

var LoadBalancingStrategies = []string{LOAD_BALANCE_RR, LOAD_BALANCE_LC}

// ServiceBrokerConfig configuration parameters
type ServiceBrokerConfig struct {
	ID               string
	Name             string
	Description      string
	DisplayName      string
	LongDescription  string
	DocumentationURL string
	SupportURL       string
	ImageURL         string
	ProviderName     string
}

var defaultServiceBrokerConfig = ServiceBrokerConfig{
	ID:               "",
	Name:             "f5servicebroker",
	Description:      "Bind F5 services to your routes",
	DisplayName:      "F5 Service Broker",
	LongDescription:  "Bind F5 services to your routes",
	DocumentationURL: "http://clouddocs.f5.com/containers/v2/cloudfoundry/",
	SupportURL:       "http://clouddocs.f5.com/containers/v2/cloudfoundry/",
	ImageURL:         "https://cdn.f5.com/websites/support/assets/images/logo.svg",
	ProviderName:     "F5 Networks",
}

type StatusConfig struct {
	Host string `yaml:"host"`
	Port uint16 `yaml:"port"`
	User string `yaml:"user"`
	Pass string `yaml:"pass"`
}

// DefaultTier2IPRange is the default tier2 virtual server IP range
var DefaultTier2IPRange = "172.0.0.0/24"

// BigIPConfig configuration parameters for bigip integration
type BigIPConfig struct {
	URL               string   `yaml:"url" json:"url"`
	User              string   `yaml:"user" json:"username"`
	Pass              string   `yaml:"pass" json:"password"`
	Partitions        []string `yaml:"partition" json:"partitions"`
	LoadBalancingMode string   `yaml:"load_balancing_mode" json:"-"`
	VerifyInterval    int      `yaml:"verify_interval" json:"-"`
	ExternalAddr      string   `yaml:"external_addr" json:"-"`
	SSLProfiles       []string `yaml:"ssl_profiles" json:"-"`
	Policies          []string `yaml:"policies" json:"-"`
	Profiles          []string `yaml:"profiles" json:"-"`
	HealthMonitors    []string `yaml:"health_monitors" json:"-"`
	DriverCmd         string   `yaml:"driver_path" json:"-"`
	Tier2IPRange      string   `yaml:"tier2_ip_range" json:"-"`
}

var defaultBigIPConfig = BigIPConfig{
	URL:               "",
	User:              "",
	Pass:              "",
	Partitions:        []string{},
	LoadBalancingMode: LoadBalancingStrategies[0],
	VerifyInterval:    30,
	ExternalAddr:      "",
	SSLProfiles:       []string{},
	Policies:          []string{},
	Profiles:          []string{},
	DriverCmd:         "",
	Tier2IPRange:      DefaultTier2IPRange,
}

var defaultStatusConfig = StatusConfig{
	Host: "0.0.0.0",
	Port: 8080,
	User: "",
	Pass: "",
}

type NatsConfig struct {
	Host string `yaml:"host"`
	Port uint16 `yaml:"port"`
	User string `yaml:"user"`
	Pass string `yaml:"pass"`
}

type RoutingApiConfig struct {
	Uri          string `yaml:"uri"`
	Port         int    `yaml:"port"`
	AuthDisabled bool   `yaml:"auth_disabled"`
}

var defaultNatsConfig = NatsConfig{
	Host: "localhost",
	Port: 4222,
	User: "",
	Pass: "",
}

type OAuthConfig struct {
	TokenEndpoint     string `yaml:"token_endpoint"`
	Port              int    `yaml:"port"`
	SkipSSLValidation bool   `yaml:"skip_ssl_validation"`
	ClientName        string `yaml:"client_name"`
	ClientSecret      string `yaml:"client_secret"`
	CACerts           string `yaml:"ca_certs"`
}

type LoggingConfig struct {
	Syslog             string `yaml:"syslog"`
	Level              string `yaml:"level"`
	LoggregatorEnabled bool   `yaml:"loggregator_enabled"`
	MetronAddress      string `yaml:"metron_address"`

	// This field is populated by the `Process` function.
	JobName string `yaml:"-"`
}

type AccessLog struct {
	File            string `yaml:"file"`
	EnableStreaming bool   `yaml:"enable_streaming"`
}

type Tracing struct {
	EnableZipkin bool `yaml:"enable_zipkin"`
}

var defaultLoggingConfig = LoggingConfig{
	Level:         "info",
	MetronAddress: "localhost:3457",
}

type Config struct {
	BigIP                    BigIPConfig         `yaml:"bigip"`
	Status                   StatusConfig        `yaml:"status"`
	Broker                   ServiceBrokerConfig `yaml:"broker"`
	Nats                     []NatsConfig        `yaml:"nats"`
	Logging                  LoggingConfig       `yaml:"logging"`
	Port                     uint16              `yaml:"port"`
	Index                    uint                `yaml:"index"`
	Zone                     string              `yaml:"zone"`
	GoMaxProcs               int                 `yaml:"go_max_procs,omitempty"`
	Tracing                  Tracing             `yaml:"tracing"`
	TraceKey                 string              `yaml:"trace_key"`
	AccessLog                AccessLog           `yaml:"access_log"`
	EnableAccessLogStreaming bool                `yaml:"enable_access_log_streaming"`
	DebugAddr                string              `yaml:"debug_addr"`
	EnablePROXY              bool                `yaml:"enable_proxy"`
	EnableSSL                bool                `yaml:"enable_ssl"`
	SSLPort                  uint16              `yaml:"ssl_port"`
	SSLCertPath              string              `yaml:"ssl_cert_path"`
	SSLKeyPath               string              `yaml:"ssl_key_path"`
	SSLCertificate           tls.Certificate
	SkipSSLValidation        bool `yaml:"skip_ssl_validation"`
	ForceForwardedProtoHttps bool `yaml:"force_forwarded_proto_https"`

	CipherString string `yaml:"cipher_suites"`
	CipherSuites []uint16

	LoadBalancerHealthyThreshold    time.Duration `yaml:"load_balancer_healthy_threshold"`
	PublishStartMessageInterval     time.Duration `yaml:"publish_start_message_interval"`
	SuspendPruningIfNatsUnavailable bool          `yaml:"suspend_pruning_if_nats_unavailable"`
	PruneStaleDropletsInterval      time.Duration `yaml:"prune_stale_droplets_interval"`
	DropletStaleThreshold           time.Duration `yaml:"droplet_stale_threshold"`
	PublishActiveAppsInterval       time.Duration `yaml:"publish_active_apps_interval"`
	StartResponseDelayInterval      time.Duration `yaml:"start_response_delay_interval"`
	EndpointTimeout                 time.Duration `yaml:"endpoint_timeout"`
	RouteServiceTimeout             time.Duration `yaml:"route_services_timeout"`
	RouteMode                       string        `yaml:"route_mode"`
	BrokerMode                      bool          `yaml:"broker_mode"`
	RoutingMode                     RoutingMode

	DrainWait          time.Duration `yaml:"drain_wait,omitempty"`
	DrainTimeout       time.Duration `yaml:"drain_timeout,omitempty"`
	SecureCookies      bool          `yaml:"secure_cookies"`
	RouterGroupName    string        `yaml:"router_group"`
	TCPRouterGroupName string        `yaml:"tcp_router_group"`

	OAuth                      OAuthConfig      `yaml:"oauth"`
	RoutingApi                 RoutingApiConfig `yaml:"routing_api"`
	RouteServiceSecret         string           `yaml:"route_services_secret"`
	RouteServiceSecretPrev     string           `yaml:"route_services_secret_decrypt_only"`
	RouteServiceRecommendHttps bool             `yaml:"route_services_recommend_https"`
	// These fields are populated by the `Process` function.
	Ip                     string        `yaml:"-"`
	RouteServiceEnabled    bool          `yaml:"-"`
	NatsClientPingInterval time.Duration `yaml:"-"`

	ExtraHeadersToLog []string `yaml:"extra_headers_to_log"`

	TokenFetcherMaxRetries                    uint32        `yaml:"token_fetcher_max_retries"`
	TokenFetcherRetryInterval                 time.Duration `yaml:"token_fetcher_retry_interval"`
	TokenFetcherExpirationBufferTimeInSeconds int64         `yaml:"token_fetcher_expiration_buffer_time"`

	PidFile     string `yaml:"pid_file"`
	LoadBalance string `yaml:"balancing_algorithm"`

	SessionPersistence bool `yaml:"session_persistence"`

	DisableKeepAlives   bool `yaml:"disable_keep_alives"`
	MaxIdleConns        int  `yaml:"max_idle_conns"`
	MaxIdleConnsPerHost int  `yaml:"max_idle_conns_per_host"`
}

var defaultConfig = Config{
	BigIP:   defaultBigIPConfig,
	Broker:  defaultServiceBrokerConfig,
	Status:  defaultStatusConfig,
	Nats:    []NatsConfig{defaultNatsConfig},
	Logging: defaultLoggingConfig,

	Port:        8081,
	Index:       0,
	GoMaxProcs:  -1,
	EnablePROXY: false,
	EnableSSL:   false,
	SSLPort:     443,

	EndpointTimeout:     60 * time.Second,
	RouteServiceTimeout: 60 * time.Second,

	PublishStartMessageInterval:               30 * time.Second,
	PruneStaleDropletsInterval:                30 * time.Second,
	DropletStaleThreshold:                     120 * time.Second,
	PublishActiveAppsInterval:                 0 * time.Second,
	StartResponseDelayInterval:                5 * time.Second,
	TokenFetcherMaxRetries:                    3,
	TokenFetcherRetryInterval:                 5 * time.Second,
	TokenFetcherExpirationBufferTimeInSeconds: 30,
	RouteMode:  HTTP.String(),
	BrokerMode: false,

	LoadBalance: LOAD_BALANCE_RR,

	SessionPersistence: true,

	DisableKeepAlives:   true,
	MaxIdleConns:        100,
	MaxIdleConnsPerHost: 2,
}

func DefaultConfig() *Config {
	c := defaultConfig
	c.Process()

	return &c
}

func (c *Config) Process() {
	var err error

	if c.GoMaxProcs == -1 {
		c.GoMaxProcs = runtime.NumCPU()
	}

	c.Logging.JobName = "cf-bigip-ctlr"
	if c.StartResponseDelayInterval > c.DropletStaleThreshold {
		c.DropletStaleThreshold = c.StartResponseDelayInterval
	}

	// To avoid routes getting purged because of unresponsive NATS server
	// we need to set the ping interval of nats client such that it fails over
	// to next NATS server before dropletstalethreshold is hit. We are hardcoding the ping interval
	// to 20 sec because the operators cannot set the value of DropletStaleThreshold and StartResponseDelayInterval
	// ping_interval = ((DropletStaleThreshold- StartResponseDelayInterval)-minimumRegistrationInterval+(2 * number_of_nats_servers))/3
	c.NatsClientPingInterval = 20 * time.Second

	if c.DrainTimeout == 0 || c.DrainTimeout == defaultConfig.EndpointTimeout {
		c.DrainTimeout = c.EndpointTimeout
	}

	c.Ip, err = localip.LocalIP()
	if err != nil {
		panic(err)
	}

	if c.EnableSSL {
		c.CipherSuites = c.processCipherSuites()
		cert, err := tls.LoadX509KeyPair(c.SSLCertPath, c.SSLKeyPath)
		if err != nil {
			panic(err)
		}
		c.SSLCertificate = cert
	}

	if c.RouteServiceSecret != "" {
		c.RouteServiceEnabled = true
	}

	// check if valid load balancing strategy
	validLb := false
	for _, lb := range LoadBalancingStrategies {
		if c.LoadBalance == lb {
			validLb = true
			break
		}
	}
	if !validLb {
		errMsg := fmt.Sprintf("Invalid load balancing algorithm %s. Allowed values are %s", c.LoadBalance, LoadBalancingStrategies)
		panic(errMsg)
	}

	if c.RouterGroupName != "" && !c.RoutingApiEnabled() {
		errMsg := fmt.Sprintf("Routing API must be enabled to assign Router Group")
		panic(errMsg)
	}

	if len(c.TCPRouterGroupName) == 0 {
		c.TCPRouterGroupName = "default-tcp"
	}

	switch c.RouteMode {
	case "http":
		c.RoutingMode = HTTP
	case "tcp":
		c.RoutingMode = TCP
	case "all":
		c.RoutingMode = all
	default:
		errMsg := fmt.Sprintf("Invalid Router Mode set %s. Allowed values are 'tcp', 'http', and 'all'.", c.RouteMode)
		panic(errMsg)
	}

	if (c.RoutingMode == all) && !c.RoutingApiEnabled() {
		c.RoutingMode = HTTP
		fmt.Print("Route mode changed from 'all' to 'HTTP', Routing API is disabled")
	}

	if (c.RoutingMode == TCP) && !c.RoutingApiEnabled() {
		errMsg := fmt.Sprintf("Routing API must be enable for TCP only route mode")
		panic(errMsg)
	}

	if c.BrokerMode && (c.Status.User == "" || c.Status.Pass == "") {
		panic("status user and pass must be set to run in service_broker mode")
	}
}

func (c *Config) processCipherSuites() []uint16 {
	cipherMap := map[string]uint16{
		"TLS_RSA_WITH_RC4_128_SHA":                0x0005,
		"TLS_RSA_WITH_3DES_EDE_CBC_SHA":           0x000a,
		"TLS_RSA_WITH_AES_128_CBC_SHA":            0x002f,
		"TLS_RSA_WITH_AES_256_CBC_SHA":            0x0035,
		"TLS_RSA_WITH_AES_128_GCM_SHA256":         0x009c,
		"TLS_RSA_WITH_AES_256_GCM_SHA384":         0x009d,
		"TLS_ECDHE_ECDSA_WITH_RC4_128_SHA":        0xc007,
		"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA":    0xc009,
		"TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA":    0xc00a,
		"TLS_ECDHE_RSA_WITH_RC4_128_SHA":          0xc011,
		"TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA":     0xc012,
		"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA":      0xc013,
		"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA":      0xc014,
		"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256":   0xc02f,
		"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": 0xc02b,
		"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384":   0xc030,
		"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384": 0xc02c}

	var ciphers []string

	if len(strings.TrimSpace(c.CipherString)) == 0 {
		panic("must specify list of cipher suite when ssl is enabled")
	} else {
		ciphers = strings.Split(c.CipherString, ":")
	}

	return convertCipherStringToInt(ciphers, cipherMap)
}

func convertCipherStringToInt(cipherStrs []string, cipherMap map[string]uint16) []uint16 {
	ciphers := []uint16{}
	for _, cipher := range cipherStrs {
		if val, ok := cipherMap[cipher]; ok {
			ciphers = append(ciphers, val)
		} else {
			var supportedCipherSuites = []string{}
			for key, _ := range cipherMap {
				supportedCipherSuites = append(supportedCipherSuites, key)
			}
			errMsg := fmt.Sprintf("Invalid cipher string configuration: %s, please choose from %v", cipher, supportedCipherSuites)
			panic(errMsg)
		}
	}

	return ciphers
}

func (c *Config) NatsServers() []string {
	var natsServers []string
	for _, info := range c.Nats {
		uri := url.URL{
			Scheme: "nats",
			User:   url.UserPassword(info.User, info.Pass),
			Host:   fmt.Sprintf("%s:%d", info.Host, info.Port),
		}
		natsServers = append(natsServers, uri.String())
	}

	return natsServers
}

func (c *Config) RoutingApiEnabled() bool {
	return (c.RoutingApi.Uri != "") && (c.RoutingApi.Port != 0)
}

func (c *Config) Initialize(configYAML []byte) error {
	c.Nats = []NatsConfig{}
	return yaml.Unmarshal(configYAML, &c)
}

func InitConfigFromFile(path string) *Config {
	var c *Config = DefaultConfig()
	var e error

	b, e := ioutil.ReadFile(path)
	if e != nil {
		panic(e.Error())
	}

	e = c.Initialize(b)
	if e != nil {
		panic(e.Error())
	}

	c.Process()

	return c
}
